server/commands/src/arguments.rs

```diff
@@ -300,3 +340,89 @@ impl AsRef<str> for TextArgument {
         self.0.as_str()
     }
 }
+
+#[derive(Debug, Error)]
+pub enum ItemParseError {
+    #[error("Unknown item {0}")]
+    ItemDoesNotExist(String),
+}
+
+#[derive(Clone, Debug)]
+pub struct ItemArgument(pub Item);
+
+impl ArgumentKind<CommandCtx> for ItemArgument {
+    type ParseError = ItemParseError;
+
+    fn satisfies<'a>(_ctx: &CommandCtx, input: &mut Input<'a>) -> bool {
+        !input.advance_until(" ").is_empty()
+    }
+
+    fn parse<'a>(_ctx: &CommandCtx, input: &mut Input<'a>) -> Result<Self, Self::ParseError> {
+        let text = input.advance_until(" ");
+        let item = Item::from_identifier(text);
+        match item {
+            Some(s) => Ok(ItemArgument(s)),
+            None => Err(ItemParseError::ItemDoesNotExist(text.to_owned())),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum I32ParseError {
+    #[error("Invalid integer {0}")]
+    Invalid(String),
+}
+
+#[derive(Clone, Debug)]
+pub struct I32Argument(pub i32);
+
+impl ArgumentKind<CommandCtx> for I32Argument {
+    type ParseError = I32ParseError;
+
+    fn satisfies<'a>(_ctx: &CommandCtx, input: &mut Input<'a>) -> bool {
+        !input.advance_until(" ").is_empty()
+    }
+
+    fn parse<'a>(_ctx: &CommandCtx, input: &mut Input<'a>) -> Result<Self, Self::ParseError> {
+        let text = input.advance_until(" ");
+        let number = text.parse::<i32>();
+        match number {
+            Ok(s) => Ok(I32Argument(s)),
+            Err(_) => Err(I32ParseError::Invalid(text.to_owned())),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum PositiveI32ParseError {
+    #[error("Invalid integer {0}")]
+    Invalid(String),
+    #[error("Integer must not be less than 0, found {0}")]
+    Negative(i32),
+}
+
+#[derive(Clone, Debug)]
+pub struct PositiveI32Argument(pub i32);
```

Comment:
Out of curiosity - why do we need a `PositiveI32Argument` instead of just a `u32`? Is it for compatibility with vanilla, which always uses signed integers?

Reply:
It is yes. Really what should be happening is that we call I32Argument but with a range argument which it checks before returning the value. Not sure how to do that

Reply:
Hmm, that would require const generics, which are unstable. This will do fine, then :)